machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -100,6 +125,325 @@ def write(self, message, event_time=0, key=None):
+            # set socket to nonblocking
+            self._conn.setblocking(0)
+            # initialize the asynchat handler
+            asynchat.async_chat.__init__(self, sock=self._conn)
+
+    def _handle_notify_ack(self, msg):
+        self.update_stream(msg.stream_id,
+                           point_of_ref = msg.point_of_ref,
+                           is_open = msg.notify_success)
+
+    def _handle_ack(self, msg):
+        self.credits += msg.credits
+        for (stream_id, point_of_ref) in msg.acks:
+            self.update_stream(stream_id,
+                               point_of_ref = point_of_ref)
+
+    def update_stream(self, stream_id, stream_name=None,
+                      point_of_ref=None, is_open=None):
+        old = self._streams.get(stream_id, None)
+        if old is None:
+            new = Stream(stream_id, stream_name, point_of_ref, is_open)
+        else:
+            if stream_name is not None:
+                if old.name != steam_name:
+                    # stream collision... throw error and close connection
+                    raise ConnectorError("Got wrong stream name for "
+                                         "stream. Expected {} but got {}."
+                                         .format(old.name, stream_name))
+            new = Stream(stream_id,
+                         old.name if old is not None else stream_name,
+                         (point_of_ref if point_of_ref is not None else
+                          old.point_of_ref),
+                         is_open if is_open is not None else old.is_open)
+        # update the stream information
+        self._streams[stream_id] = new
+        # Call optional user handler with updated stream
+        self.stream_updated(new)
+
+    ##########################
+    # Outoing communications #
+    ##########################
+
+    def initiate_handshake(self, hello):
+        conn = socket.socket()
+        conn.connect( (self._host, self._port) )
+        self._conn = conn
+
+        self.in_handshake = True
+        self.version = hello.version
+        self.cookie = hello.cookie
+        self.program_name = hello.program_name
+        self.instance_name = hello.instance_name
+        self._conn.setblocking(1) # Set socket to blocking mode
+        self._conn.sendall(cwm.Frame.encode(hello))
+        header_bytes = self._conn.recv(4)
+        frame_size = struct.unpack('>I', header_bytes)[0]
+        frame = self._conn.recv(frame_size)
+        try:
+            self._handle_frame(frame)
+        except Exception as err:
+            # close the connection and raise the error
+            self._conn.close()
+            raise err
+
+    def write(self, msg):
+        if not self.is_alive():
+            raise ConnectorError("Can't write to a closed conection")
+        if (isinstance(msg, (cwm.Error, cwm.Notify)) or
+            (msg.stream_id in self._streams and
+            self._streams[msg.stream_id].is_open)):
```

Comment:
Should a user be allowed to send a message to an open/not-yet-existing stream_id?
(e.g. could the library implicitly add the notify?)

Reply:
(from @jtfmumm ): Ideally like to see:
- User can optimistically continue sending messages without having to worry about the state machien of the protocol
- They may be notified through our API that it's time to rewind

(from @nisanharamati):
- What happens when you queue messages with higher point of references, before you finally get the (asynchronous) rewind instruction?
    - @jtfmumm : from the wallaroo pony side, as long as we wait for the exact point of ref requested in the rewind
    - @strmpnk : we can use a boundary message to indiciate this.
        - @nisanharamati : maybe we can add a "rewound to X" message type
    - @jtfmumm : during a rollback, can stay in "rollback state" until a rewind seems to have completed successfully, however we choose to indiciate that.

Reply:
I think notify is a protocol level concern and not an API one. Either could work for the lower level event callbacks but I think it makes more sense to handle notifications implicitly.

Reply:
The biggest dificulty in implicit notify is that the current protocol requires synchrony in notify->notify_ack.
There are several ways to approach this:
1. We could send messages and rely on Wallaroo to discard them.
2. We could buffer messages in a special `awaiting_ack` queue (per stream)
3. We could call out to user defined functions `stream_opened` and `stream_closed`
4. [this is what the current code does] We could raise an error when trying to send a message on a non-open stream (including non-existing stream, since both cases require the same notify_ack to open)

All of these will work, but we should settle on one and move forward with it.

My preference for a first implementation is towards the more "raw" version in (4), which we can polish and improve on in a later update.